python/ray/util/collective/collective_group/hccl_collective_group.py

```diff
@@ -0,0 +1,816 @@
+        self.root_info_bytes = root_info_bytes
+        return self.root_info_bytes
+
+    def get_root_info_bytes(self):
+        if not self.root_info_bytes:
+            logger.warning(
+                f"The HcclRootInfo has not been set yet for store {self.name}."
+            )
+        return self.root_info_bytes
+
+
+class HCCLGroup(BaseGroup):
+    def __init__(self, world_size, rank, group_name):
+        """Init an HCCL collective group."""
+        # Ensure HCCL backend is registered with torch.distributed
+        import torch_npu  # noqa: F401
+
+        metadata_key = get_master_address_metadata_key(group_name)
+        try:
+            metadata = internal_kv._internal_kv_get(metadata_key)
+        except ValueError:
+            raise RuntimeError(
+                f"HCCLGroup expected metadata in internal_kv with name `{metadata_key}`. "
+                "HCCLGroup should not be instantiated directly. "
+                "Use ray.experimental.collective.create_collective_group to create the group."
+            )
+
+        metadata = metadata.decode()
+        master_addr, master_port = metadata.split(":")
+        os.environ["MASTER_ADDR"] = master_addr
+        os.environ["MASTER_PORT"] = master_port
+        dist.init_process_group(backend="hccl", rank=rank, world_size=world_size)
+        super(HCCLGroup, self).__init__(world_size, rank, group_name)
+        self._dev_comm_map = {}
+        self._dev_streams_map = {}
+
+        # record used NPU IDs.
+        self._used_npu_indices = set()
+        self.libhccl = libhccl
+
+    def destroy_group(self):
+        if len(self._dev_comm_map.keys()) > 0:
+            for comm_key, comms in self._dev_comm_map.items():
+                if comms:
+                    for comm in comms:
+                        self.libhccl.HcclCommDestroy(comm)
+                self._dev_comm_map[comm_key] = None
+
+        if self.rank == 0:
+            for comm_key in self._dev_comm_map:
+                assert not self._dev_comm_map[comm_key]
+                group_key = self._generate_group_key(comm_key)
+                self._destroy_store(group_key)
+        self._barrier_tensor = None
+        self._dev_comm_map = None
+        self._dev_streams_map = None
+        super(HCCLGroup, self).destroy_group()
+
+    @classmethod
+    def backend(cls):
+        return Backend.HCCL
+
+    def broadcast(self, tensors, broadcast_options=BroadcastOptions()):
+        """Broadcast tensors to all other npus following options.
+
+        Args:
+            tensors: tensors to be broadcast or received.
+            broadcast_options: broadcast options.
+
+        Returns:
+            None
+        """
+        root_rank = (
+            len(tensors) * broadcast_options.root_rank + broadcast_options.root_tensor
+        )
+
+        def collective_fn(
+            input_tensor: torch.Tensor, output_tensor: torch.Tensor, comm, stream
+        ):
+            with torch.npu.device(input_tensor.device):
```

Comment:
If this backend only supports torch.Tensor, please add checks for that, similar to what we do for [torch_gloo](https://github.com/ray-project/ray/blob/ec20f9a16ef07cc82a74bac72b47abdc27fd6e5c/python/ray/util/collective/collective_group/torch_gloo_collective_group.py#L141).